File: exporters.py
"""
exporters.py — CSV and Excel export for timetables and master data.

The app could print and publish HTML but had no way to get data OUT into a
spreadsheet, which is what schools actually pass around. This covers:

  * timetable by class, by teacher, by room
  * lesson (assignment) list
  * teacher / class / room / subject lists

CSV is written with a UTF-8 BOM and ';' separator, because that is what Excel on a
Turkish Windows install opens correctly by double-click — a plain comma-separated
UTF-8 file shows up as one mangled column, which users read as "the export is
broken".

Excel output needs openpyxl. When it is missing the caller is told so rather than
silently getting a CSV with an .xlsx extension, which Excel refuses to open.
"""
import csv
import os
from collections import defaultdict

try:
    from openpyxl import Workbook
    from openpyxl.styles import Alignment, Border, Font, PatternFill, Side
    from openpyxl.utils import get_column_letter
    HAS_XLSX = True
except ImportError:      # pragma: no cover - depends on the install
    HAS_XLSX = False


CSV_DELIMITER = ";"
CSV_ENCODING = "utf-8-sig"   # the BOM is what makes Excel detect UTF-8


# ── Shared helpers ───────────────────────────────────────────────────

def _norm(value) -> str:
    return " ".join(str(value or "").split()).strip()


def grid_dimensions(data_store: dict):
    settings = data_store.get("settings", {}) or {}
    days = settings.get("days")
    if not days:
        count = int(settings.get("day_count", data_store.get("gun_sayisi", 5)))
        days = ["Pazartesi", "Salı", "Çarşamba", "Perşembe", "Cuma",
                "Cumartesi", "Pazar"][:count]
    periods = int(settings.get("periods", data_store.get("ders_saati", 8))) or 8
    return list(days), periods


def period_labels(data_store: dict, periods: int):
    """Column headers, using the bell schedule when the school has defined one."""
    bells = data_store.get("zil_programi") or data_store.get("bell_times") or {}
    labels = []
    for p in range(periods):
        entry = None
        if isinstance(bells, dict):
            entry = bells.get(str(p)) or bells.get(p)
        elif isinstance(bells, list) and p < len(bells):
            entry = bells[p]
        if isinstance(entry, dict):
            start = entry.get("start") or entry.get("baslangic") or ""
            end = entry.get("end") or entry.get("bitis") or ""
            if start and end:
                labels.append(f"{p + 1}. Ders\n{start}-{end}")
                continue
        labels.append(f"{p + 1}. Ders")
    return labels


def _placement_index(data_store: dict):
    """(entity, day, period) -> placement, for each of the three views."""
    by_class = {}
    by_teacher = {}
    by_room = {}
    for p in data_store.get("grid_placements", []) or []:
        if not isinstance(p, dict):
            continue
        day = int(p.get("day", p.get("col", 0)) or 0)
        start = int(p.get("period", p.get("row", 0)) or 0)
        span = max(1, int(p.get("duration", 1) or 1))
        cls = _norm(p.get("class_name") or p.get("class"))
        teacher = _norm(p.get("teacher_name") or p.get("teacher"))
        room = _norm(p.get("room_name") or p.get("room") or p.get("derslik"))
        for off in range(span):
            slot = (day, start + off)
            if cls:
                by_class[(cls, *slot)] = p
            if teacher:
                by_teacher[(teacher, *slot)] = p
            if room:
                by_room[(room, *slot)] = p
    return by_class, by_teacher, by_room


def _cell_text(placement: dict, view: str) -> str:
    if not placement:
        return ""
    subject = _norm(placement.get("subject_name") or placement.get("subject"))
    teacher = _norm(placement.get("teacher_name") or placement.get("teacher"))
    cls = _norm(placement.get("class_name") or placement.get("class"))
    room = _norm(placement.get("room_name") or placement.get("room") or placement.get("derslik"))

    if view == "classes":
        parts = [subject, teacher]
    elif view == "teachers":
        parts = [subject, cls]
    else:                      # rooms
        parts = [subject, cls, teacher]
    if room and view != "rooms":
        parts.append(room)
    return "\n".join(x for x in parts if x)


def build_timetable_table(data_store: dict, view: str = "classes"):
    """(headers, rows) for a printable timetable grid.

    view: "classes" | "teachers" | "rooms"
    """
    days, periods = grid_dimensions(data_store)
    labels = period_labels(data_store, periods)
    by_class, by_teacher, by_room = _placement_index(data_store)

    if view == "teachers":
        entities = [_norm(t.get("ad") or t.get("name"))
                    for t in data_store.get("ogretmenler", []) or [] if isinstance(t, dict)]
        index, first_col = by_teacher, "Öğretmen"
    elif view == "rooms":
        entities = [_norm(r.get("ad") or r.get("name"))
                    for r in data_store.get("derslikler", []) or [] if isinstance(r, dict)]
        index, first_col = by_room, "Derslik"
    else:
        entities = [_norm(c.get("ad") or c.get("name"))
                    for c in data_store.get("siniflar", []) or [] if isinstance(c, dict)]
        index, first_col = by_class, "Sınıf"

    entities = [e for e in entities if e]

    headers = [first_col]
    for day in days:
        for lbl in labels:
            headers.append(f"{day} {lbl.splitlines()[0]}")

    rows = []
    for entity in entities:
        row = [entity]
        for d in range(len(days)):
            for p in range(periods):
                row.append(_cell_text(index.get((entity, d, p)), view))
        rows.append(row)
    return headers, rows


def build_lesson_list(data_store: dict):
    headers = ["Sınıf", "Ders", "Öğretmen", "Derslik", "Haftalık Saat", "Dağılım"]
    rows = []
    for a in data_store.get("atamalar", []) or []:
        if not isinstance(a, dict):
            continue
        rows.append([
            _norm(a.get("class") or a.get("sinif") or a.get("class_name")),
            _norm(a.get("subject") or a.get("ders")),
            _norm(a.get("teacher") or a.get("ogretmen") or a.get("teacher_name")),
            _norm(a.get("room") or a.get("derslik")),
            a.get("duration") or a.get("saat") or "",
            _norm(a.get("type") or a.get("dagilim")),
        ])
    return headers, rows


def build_entity_list(data_store: dict, kind: str):
    """kind: 'ogretmenler' | 'siniflar' | 'derslikler' | 'dersler'"""
    specs = {
        "ogretmenler": (["Ad", "Kısa Ad", "Branş", "Günlük Max", "Haftalık Max"],
                        ["ad", "kisa", "brans", "max_gunluk", "max_haftalik"]),
        "siniflar":    (["Ad", "Kısa Ad", "Mevcut", "Sınıf Öğretmeni"],
                        ["ad", "kisa", "mevcut", "sinif_ogretmeni"]),
        "derslikler":  (["Ad", "Kısa Ad", "Kapasite", "Bina", "Tür"],
                        ["ad", "kisa", "kapasite", "bina", "tur"]),
        "dersler":     (["Ad", "Kısa Ad", "Renk"], ["ad", "kisa", "renk"]),
    }
    headers, fields = specs.get(kind, (["Ad"], ["ad"]))
    rows = []
    for item in data_store.get(kind, []) or []:
        if isinstance(item, dict):
            rows.append([_norm(item.get(f, "")) for f in fields])
    return headers, rows


# ── Writers ──────────────────────────────────────────────────────────

def write_csv(path: str, headers, rows) -> str:
    os.makedirs(os.path.dirname(os.path.abspath(path)) or ".", exist_ok=True)
    with open(path, "w", encoding=CSV_ENCODING, newline="") as f:
        writer = csv.writer(f, delimiter=CSV_DELIMITER, quoting=csv.QUOTE_MINIMAL)
        writer.writerow(headers)
        for row in rows:
            # Newlines inside a cell survive the quoting, but Excel renders them only
            # with wrap-on; flatten to " / " so the CSV stays readable either way.
            writer.writerow([str(c).replace("\n", " / ") if c else "" for c in row])
    return path


def write_xlsx(path: str, sheets) -> str:
    """sheets: [(sheet_name, headers, rows), ...]"""
    if not HAS_XLSX:
        raise RuntimeError(
            "Excel çıktısı için 'openpyxl' paketi gerekli. "
            "CSV olarak kaydedebilir veya 'pip install openpyxl' çalıştırabilirsiniz.")

    os.makedirs(os.path.dirname(os.path.abspath(path)) or ".", exist_ok=True)
    wb = Workbook()
    wb.remove(wb.active)

    head_fill = PatternFill("solid", fgColor="1F4E79")
    head_font = Font(bold=True, color="FFFFFF", size=10)
    body_font = Font(size=9)
    thin = Side(style="thin", color="B0B0B0")
    border = Border(left=thin, right=thin, top=thin, bottom=thin)
    centered = Alignment(horizontal="center", vertical="center", wrap_text=True)
    left = Alignment(horizontal="left", vertical="center", wrap_text=True)

    for name, headers, rows in sheets:
        # Excel rejects >31 chars and the characters below in a sheet name.
        safe = str(name)[:31]
        for ch in "[]:*?/\\":
            safe = safe.replace(ch, "-")
        ws = wb.create_sheet(safe or "Sayfa")

        ws.append(list(headers))
        for cell in ws[1]:
            cell.fill = head_fill
            cell.font = head_font
            cell.alignment = centered
            cell.border = border

        for row in rows:
            ws.append(list(row))

        widths = [len(str(h)) for h in headers]
        for row in rows:
            for i, value in enumerate(row):
                if i < len(widths):
                    longest = max((len(part) for part in str(value or "").split("\n")), default=0)
                    widths[i] = max(widths[i], longest)
        for i, width in enumerate(widths, start=1):
            ws.column_dimensions[get_column_letter(i)].width = min(max(width + 2, 9), 34)

        for row in ws.iter_rows(min_row=2):
            for cell in row:
                cell.font = body_font
                cell.border = border
                cell.alignment = left if cell.column == 1 else centered

        ws.freeze_panes = "B2"          # keep the entity column and header visible
        ws.print_options.horizontalCentered = True
        ws.page_setup.orientation = "landscape"
        ws.page_setup.fitToWidth = 1

    wb.save(path)
    return path


# ── High-level entry points ──────────────────────────────────────────

EXPORT_KINDS = [
    ("timetable_classes", "Ders Programı — Sınıflara Göre"),
    ("timetable_teachers", "Ders Programı — Öğretmenlere Göre"),
    ("timetable_rooms", "Ders Programı — Dersliklere Göre"),
    ("lessons", "Ders Atama Listesi"),
    ("teachers", "Öğretmen Listesi"),
    ("classes", "Sınıf Listesi"),
    ("rooms", "Derslik Listesi"),
    ("subjects", "Ders Listesi"),
]


def build_sheet(data_store: dict, kind: str):
    """(sheet_title, headers, rows) for one export kind."""
    if kind == "timetable_classes":
        return "Sınıf Programı", *build_timetable_table(data_store, "classes")
    if kind == "timetable_teachers":
        return "Öğretmen Programı", *build_timetable_table(data_store, "teachers")
    if kind == "timetable_rooms":
        return "Derslik Programı", *build_timetable_table(data_store, "rooms")
    if kind == "lessons":
        return "Ders Atamaları", *build_lesson_list(data_store)
    if kind == "teachers":
        return "Öğretmenler", *build_entity_list(data_store, "ogretmenler")
    if kind == "classes":
        return "Sınıflar", *build_entity_list(data_store, "siniflar")
    if kind == "rooms":
        return "Derslikler", *build_entity_list(data_store, "derslikler")
    if kind == "subjects":
        return "Dersler", *build_entity_list(data_store, "dersler")
    raise ValueError(f"bilinmeyen dışa aktarım türü: {kind}")


def export_to_file(data_store: dict, path: str, kinds=None) -> str:
    """Writes the chosen sections to `path`; format follows the extension.

    CSV holds a single section (the format has no concept of sheets), so the first
    selected kind is written and the caller is expected to have said so.
    """
    kinds = list(kinds or ["timetable_classes"])
    ext = os.path.splitext(path)[1].lower()

    if ext in (".xlsx", ".xlsm"):
        sheets = [build_sheet(data_store, k) for k in kinds]
        return write_xlsx(path, sheets)

    _title, headers, rows = build_sheet(data_store, kinds[0])
    return write_csv(path, headers, rows)

File: web/src/app/exporters.ts
// exporters.py karşılığı: dışa aktarılacak tabloları üretir ve .xlsx / .csv yazar.
// .xlsx için dış kütüphane yok: küçük bir ZIP (sıkıştırmasız) + OOXML yazıcısı.
import { bellTimes } from './print/reports.tsx';

const norm = (v: unknown) => String(v ?? '').split(/\s+/).filter(Boolean).join(' ').trim();
const int0 = (v: unknown) => { const n = Number(v); return Number.isFinite(n) ? Math.trunc(n) : 0; };

function gridDims(d: any): { days: string[]; periods: number } {
  const st = d.settings ?? {};
  let days: string[] = st.days;
  if (!Array.isArray(days) || !days.length) days = ['Pazartesi', 'Salı', 'Çarşamba', 'Perşembe', 'Cuma', 'Cumartesi', 'Pazar'].slice(0, int0(st.day_count ?? d.gun_sayisi ?? 5) || 5);
  return { days: [...days], periods: int0(st.periods ?? d.ders_saati ?? 8) || 8 };
}

function periodLabels(d: any, periods: number): string[] {
  const t = bellTimes(d, periods, '-');
  return Array.from({ length: periods }, (_, p) => `${p + 1}. Ders\n${t[p]}`);
}

function index(d: any) {
  const byClass = new Map<string, any>(), byTeacher = new Map<string, any>(), byRoom = new Map<string, any>();
  for (const p of d.grid_placements ?? []) {
    if (!p || typeof p !== 'object') continue;
    const day = int0(p.day ?? p.col), st = int0(p.period ?? p.row), span = Math.max(1, int0(p.duration) || 1);
    const c = norm(p.class_name || p.class), t = norm(p.teacher_name || p.teacher), r = norm(p.room_name || p.room || p.derslik);
    for (let o = 0; o < span; o++) {
      const s = `${day}|${st + o}`;
      if (c) byClass.set(`${c}|${s}`, p);
      if (t) byTeacher.set(`${t}|${s}`, p);
      if (r) byRoom.set(`${r}|${s}`, p);
    }
  }
  return { byClass, byTeacher, byRoom };
}

function cellText(p: any, view: string): string {
  if (!p) return '';
  const s = norm(p.subject_name || p.subject), t = norm(p.teacher_name || p.teacher), c = norm(p.class_name || p.class), r = norm(p.room_name || p.room || p.derslik);
  const parts = view === 'classes' ? [s, t] : view === 'teachers' ? [s, c] : [s, c, t];
  if (r && view !== 'rooms') parts.push(r);
  return parts.filter(Boolean).join('\n');
}

type Table = [string[], (string | number)[][]];

function timetable(d: any, view: 'classes' | 'teachers' | 'rooms'): Table {
  const { days, periods } = gridDims(d);
  const labels = periodLabels(d, periods);
  const ix = index(d);
  const [list, idx, first] = view === 'teachers' ? [d.ogretmenler, ix.byTeacher, 'Öğretmen'] : view === 'rooms' ? [d.derslikler, ix.byRoom, 'Derslik'] : [d.siniflar, ix.byClass, 'Sınıf'];
  const ents = (list ?? []).filter((x: any) => x && typeof x === 'object').map((x: any) => norm(x.ad || x.name)).filter(Boolean) as string[];
  const headers = [first];
  for (const day of days) for (const l of labels) headers.push(`${day} ${l.split('\n')[0]}`);
  const rows = ents.map((e) => {
    const row: string[] = [e];
    for (let dd = 0; dd < days.length; dd++) for (let p = 0; p < periods; p++) row.push(cellText(idx.get(`${e}|${dd}|${p}`), view));
    return row;
  });
  return [headers, rows];
}

function lessonList(d: any): Table {
  const rows = (d.atamalar ?? []).filter((a: any) => a && typeof a === 'object').map((a: any) => [
    norm(a.class || a.sinif || a.class_name), norm(a.subject || a.ders), norm(a.teacher || a.ogretmen || a.teacher_name), norm(a.room || a.derslik), a.duration || a.saat || '', norm(a.type || a.dagilim),
  ]);
  return [['Sınıf', 'Ders', 'Öğretmen', 'Derslik', 'Haftalık Saat', 'Dağılım'], rows];
}

function entityList(d: any, kind: string): Table {
  const specs: Record<string, [string[], string[]]> = {
    ogretmenler: [['Ad', 'Kısa Ad', 'Branş', 'Günlük Max', 'Haftalık Max'], ['ad', 'kisa', 'brans', 'max_gunluk', 'max_haftalik']],
    siniflar: [['Ad', 'Kısa Ad', 'Mevcut', 'Sınıf Öğretmeni'], ['ad', 'kisa', 'mevcut', 'sinif_ogretmeni']],
    derslikler: [['Ad', 'Kısa Ad', 'Kapasite', 'Bina', 'Tür'], ['ad', 'kisa', 'kapasite', 'bina', 'tur']],
    dersler: [['Ad', 'Kısa Ad', 'Renk'], ['ad', 'kisa', 'renk']],
  };
  const [h, f] = specs[kind] ?? [['Ad'], ['ad']];
  return [h, (d[kind] ?? []).filter((x: any) => x && typeof x === 'object').map((x: any) => f.map((k) => norm(x[k] ?? '')))];
}

export const EXPORT_KINDS: [string, string][] = [
  ['timetable_classes', 'Ders Programı — Sınıflara Göre'],
  ['timetable_teachers', 'Ders Programı — Öğretmenlere Göre'],
  ['timetable_rooms', 'Ders Programı — Dersliklere Göre'],
  ['lessons', 'Ders Atama Listesi'],
  ['teachers', 'Öğretmen Listesi'],
  ['classes', 'Sınıf Listesi'],
  ['rooms', 'Derslik Listesi'],
  ['subjects', 'Ders Listesi'],
];

export function buildSheet(d: any, kind: string): [string, string[], (string | number)[][]] {
  switch (kind) {
    case 'timetable_classes': return ['Sınıf Programı', ...timetable(d, 'classes')];
    case 'timetable_teachers': return ['Öğretmen Programı', ...timetable(d, 'teachers')];
    case 'timetable_rooms': return ['Derslik Programı', ...timetable(d, 'rooms')];
    case 'lessons': return ['Ders Atamaları', ...lessonList(d)];
    case 'teachers': return ['Öğretmenler', ...entityList(d, 'ogretmenler')];
    case 'classes': return ['Sınıflar', ...entityList(d, 'siniflar')];
    case 'rooms': return ['Derslikler', ...entityList(d, 'derslikler')];
    case 'subjects': return ['Dersler', ...entityList(d, 'dersler')];
  }
  throw new Error(`bilinmeyen dışa aktarım türü: ${kind}`);
}

// ── CSV ──────────────────────────────────────────────────────────────────
export function toCsv(headers: string[], rows: (string | number)[][]): Blob {
  const q = (v: unknown) => {
    const s = v === null || v === undefined || v === '' ? '' : String(v).replace(/\n/g, ' / ');
    return /[";\n\r]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
  };
  const text = [headers, ...rows].map((r) => r.map(q).join(';')).join('\r\n') + '\r\n';
  return new Blob(['﻿' + text], { type: 'text/csv;charset=utf-8' });
}

// ── XLSX ─────────────────────────────────────────────────────────────────
const CRC_TABLE = (() => { const t = new Uint32Array(256); for (let n = 0; n < 256; n++) { let c = n; for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1; t[n] = c >>> 0; } return t; })();
function crc32(b: Uint8Array): number { let c = 0xffffffff; for (let i = 0; i < b.length; i++) c = CRC_TABLE[(c ^ b[i]) & 0xff] ^ (c >>> 8); return (c ^ 0xffffffff) >>> 0; }

/** Sıkıştırmasız (STORE) ZIP. */
function zip(files: [string, string][]): Blob {
  const enc = new TextEncoder();
  const parts: Uint8Array[] = [], central: Uint8Array[] = [];
  let offset = 0;
  for (const [name, content] of files) {
    const data = enc.encode(content), nm = enc.encode(name), crc = crc32(data);
    const lh = new DataView(new ArrayBuffer(30));
    lh.setUint32(0, 0x04034b50, true); lh.setUint16(4, 20, true); lh.setUint16(6, 0x0800, true); lh.setUint16(8, 0, true);
    lh.setUint16(10, 0, true); lh.setUint16(12, 0x21, true); lh.setUint32(14, crc, true); lh.setUint32(18, data.length, true);
    lh.setUint32(22, data.length, true); lh.setUint16(26, nm.length, true); lh.setUint16(28, 0, true);
    parts.push(new Uint8Array(lh.buffer), nm, data);
    const ch = new DataView(new ArrayBuffer(46));
    ch.setUint32(0, 0x02014b50, true); ch.setUint16(4, 20, true); ch.setUint16(6, 20, true); ch.setUint16(8, 0x0800, true); ch.setUint16(10, 0, true);
    ch.setUint16(12, 0, true); ch.setUint16(14, 0x21, true); ch.setUint32(16, crc, true); ch.setUint32(20, data.length, true); ch.setUint32(24, data.length, true);
    ch.setUint16(28, nm.length, true); ch.setUint16(30, 0, true); ch.setUint16(32, 0, true); ch.setUint16(34, 0, true); ch.setUint16(36, 0, true);
    ch.setUint32(38, 0, true); ch.setUint32(42, offset, true);
    central.push(new Uint8Array(ch.buffer), nm);
    offset += 30 + nm.length + data.length;
  }
  const cdSize = central.reduce((s, x) => s + x.length, 0);
  const end = new DataView(new ArrayBuffer(22));
  end.setUint32(0, 0x06054b50, true); end.setUint16(8, files.length, true); end.setUint16(10, files.length, true);
  end.setUint32(12, cdSize, true); end.setUint32(16, offset, true);
  return new Blob([...parts, ...central, new Uint8Array(end.buffer)] as BlobPart[], { type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' });
}

const esc = (s: string) => s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;').replace(/[\u0000-\u0008\u000B\u000C\u000E-\u001F]/g, '');
function colName(i: number): string { let s = ''; i++; while (i > 0) { const m = (i - 1) % 26; s = String.fromCharCode(65 + m) + s; i = Math.floor((i - 1) / 26); } return s; }

export function toXlsx(sheets: [string, string[], (string | number)[][]][]): Blob {
  const files: [string, string][] = [];
  const names: string[] = [];
  sheets.forEach(([name, headers, rows], si) => {
    let safe = String(name).slice(0, 31);
    for (const ch of '[]:*?/\\') safe = safe.split(ch).join('-');
    safe = safe || 'Sayfa';
    while (names.includes(safe)) safe = `${safe.slice(0, 28)}_${si + 1}`;
    names.push(safe);
    const widths = headers.map((h) => String(h).length);
    for (const r of rows) r.forEach((v, i) => { if (i < widths.length) widths[i] = Math.max(widths[i], ...String(v ?? '').split('\n').map((p) => p.length)); });
    const cols = widths.map((w, i) => `<col min="${i + 1}" max="${i + 1}" width="${Math.min(Math.max(w + 2, 9), 34)}" customWidth="1"/>`).join('');
    const cell = (v: string | number, ref: string, style: number) => (typeof v === 'number'
      ? `<c r="${ref}" s="${style}"><v>${v}</v></c>`
      : v === '' || v === null || v === undefined ? `<c r="${ref}" s="${style}"/>` : `<c r="${ref}" s="${style}" t="inlineStr"><is><t xml:space="preserve">${esc(String(v))}</t></is></c>`);
    const xmlRows = [
      `<row r="1">${headers.map((h, i) => cell(String(h), `${colName(i)}1`, 1)).join('')}</row>`,
      ...rows.map((r, ri) => `<row r="${ri + 2}">${r.map((v, i) => cell(typeof v === 'number' ? v : String(v ?? ''), `${colName(i)}${ri + 2}`, i === 0 ? 2 : 3)).join('')}</row>`),
    ].join('');
    files.push([`xl/worksheets/sheet${si + 1}.xml`,
      `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><sheetPr><pageSetUpPr fitToPage="1"/></sheetPr>`
      + `<sheetViews><sheetView workbookViewId="0"><pane xSplit="1" ySplit="1" topLeftCell="B2" activePane="bottomRight" state="frozen"/></sheetView></sheetViews>`
      + `<cols>${cols}</cols><sheetData>${xmlRows}</sheetData><printOptions horizontalCentered="1"/>`
      + `<pageMargins left="0.4" right="0.4" top="0.5" bottom="0.5" header="0.3" footer="0.3"/><pageSetup orientation="landscape" fitToWidth="1" fitToHeight="0"/></worksheet>`]);
  });
  files.unshift(
    ['[Content_Types].xml', `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types"><Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/><Default Extension="xml" ContentType="application/xml"/><Override PartName="/xl/workbook.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml"/><Override PartName="/xl/styles.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.styles+xml"/>${names.map((_, i) => `<Override PartName="/xl/worksheets/sheet${i + 1}.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml"/>`).join('')}</Types>`],
    ['_rels/.rels', `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships"><Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" Target="xl/workbook.xml"/></Relationships>`],
    ['xl/workbook.xml', `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><workbook xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"><sheets>${names.map((n, i) => `<sheet name="${esc(n)}" sheetId="${i + 1}" r:id="rId${i + 1}"/>`).join('')}</sheets></workbook>`],
    ['xl/_rels/workbook.xml.rels', `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">${names.map((_, i) => `<Relationship Id="rId${i + 1}" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet" Target="worksheets/sheet${i + 1}.xml"/>`).join('')}<Relationship Id="rId${names.length + 1}" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles" Target="styles.xml"/></Relationships>`],
    ['xl/styles.xml', `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><styleSheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main">`
      + `<fonts count="3"><font><sz val="11"/><name val="Calibri"/></font><font><b/><sz val="10"/><color rgb="FFFFFFFF"/><name val="Calibri"/></font><font><sz val="9"/><name val="Calibri"/></font></fonts>`
      + `<fills count="3"><fill><patternFill patternType="none"/></fill><fill><patternFill patternType="gray125"/></fill><fill><patternFill patternType="solid"><fgColor rgb="FF1F4E79"/><bgColor indexed="64"/></patternFill></fill></fills>`
      + `<borders count="2"><border><left/><right/><top/><bottom/><diagonal/></border><border><left style="thin"><color rgb="FFB0B0B0"/></left><right style="thin"><color rgb="FFB0B0B0"/></right><top style="thin"><color rgb="FFB0B0B0"/></top><bottom style="thin"><color rgb="FFB0B0B0"/></bottom><diagonal/></border></borders>`
      + `<cellStyleXfs count="1"><xf numFmtId="0" fontId="0" fillId="0" borderId="0"/></cellStyleXfs>`
      + `<cellXfs count="4"><xf numFmtId="0" fontId="0" fillId="0" borderId="0" xfId="0"/>`
      + `<xf numFmtId="0" fontId="1" fillId="2" borderId="1" xfId="0" applyFont="1" applyFill="1" applyBorder="1" applyAlignment="1"><alignment horizontal="center" vertical="center" wrapText="1"/></xf>`
      + `<xf numFmtId="0" fontId="2" fillId="0" borderId="1" xfId="0" applyFont="1" applyBorder="1" applyAlignment="1"><alignment horizontal="left" vertical="center" wrapText="1"/></xf>`
      + `<xf numFmtId="0" fontId="2" fillId="0" borderId="1" xfId="0" applyFont="1" applyBorder="1" applyAlignment="1"><alignment horizontal="center" vertical="center" wrapText="1"/></xf></cellXfs>`
      + `<cellStyles count="1"><cellStyle name="Normal" xfId="0" builtinId="0"/></cellStyles></styleSheet>`],
  );
  return zip(files);
}

export function downloadBlob(blob: Blob, name: string) {
  const a = document.createElement('a');
  a.href = URL.createObjectURL(blob);
  a.download = name;
  a.click();
  setTimeout(() => URL.revokeObjectURL(a.href), 3000);
}

/** exporters.export_to_file — biçim uzantıdan; CSV tek bölüm tutar. */
export function exportFile(d: any, kinds: string[], xlsx: boolean): { blob: Blob; ext: string } {
  if (xlsx) return { blob: toXlsx(kinds.map((k) => buildSheet(d, k))), ext: '.xlsx' };
  const [, h, r] = buildSheet(d, kinds[0]);
  return { blob: toCsv(h, r), ext: '.csv' };
}
